go-controller/pkg/util/kube.go

```diff
@@ -719,55 +733,134 @@ func NoHostSubnet(node *v1.Node) bool {
+// and then iterates through them to apply the given function fn to each eligible endpoint.
+func ForEachEligibleEndpoint(endpoints []discovery.Endpoint, service *kapi.Service, fn func(ep discovery.Endpoint)) {
+	forEachSelectedAndEligibleEndpoint(endpoints, service, nil, fn)
+}
+
+// forEachSelectedAndEligibleEndpoint does the following:
+// (1) filters the given endpoints by the given condition function condFn;
+// (2) selects the eligible endpoints
+// (3) applies the provided function fn to them.
+// Eligible endpoints are ready endpoints. If there are none, eligible endpoints are serving & terminating endpoints
+// (an endpoint that is terminating shows ready=false).
+// Returns true when elegible endpoints where ready endpoints or false if they were terminating endpoints
+// The endpoint selection in step 2 above implements KEP-1669
+// (https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1669-proxy-terminating-endpoints/README.md)
+// The corresponding service needs to provided as an input argument because if Spec.PublishNotReadyAddresses is set,
+// then all provided endpoints must always be returned. PublishNotReadyAddresses tells endpoint consumers to disregard
+// any indications of ready/not-ready and is generally used together with headless services so that DNS records of
+// all endpoints (ready or not) are always published.
+func forEachSelectedAndEligibleEndpoint(endpoints []discovery.Endpoint, service *kapi.Service, condFn func(ep discovery.Endpoint) bool, fn func(ep discovery.Endpoint)) bool {
 	includeTerminating := service != nil && service.Spec.PublishNotReadyAddresses
-	var shortcut bool
-	for _, endpoint := range endpointSlice.Endpoints {
-		if IsEndpointEligible(endpoint, includeTerminating) {
-			fn(endpoint, &shortcut)
-			if shortcut {
-				// shortcircuit the whole iteration
-				return
+	var selectedEndpoints []discovery.Endpoint
+	var eligibleSelectedEndpoints []discovery.Endpoint
+	var fallbackToTerminatingEndpoints bool
+
+	// Apply precondition on endpoints, if provided
+	if condFn != nil {
+		for _, endpoint := range endpoints {
+			if condFn(endpoint) {
+				selectedEndpoints = append(selectedEndpoints, endpoint)
 			}
 		}
+	} else {
+		selectedEndpoints = endpoints
 	}
+
+	// Apply endpoint selection based on readiness
+	for _, endpoint := range selectedEndpoints {
+		if includeTerminating || IsEndpointReady(endpoint) {
+			eligibleSelectedEndpoints = append(eligibleSelectedEndpoints, endpoint)
+		}
+	}
+	serviceStr := ""
+	if service != nil {
+		serviceStr = fmt.Sprintf(" (service %s/%s)", service.Namespace, service.Name)
+	}
+	klog.V(5).Infof("Endpoint selection%s: found %d ready endpoints", serviceStr, len(eligibleSelectedEndpoints))
+
+	// Fallback to serving terminating endpoints (ready=false, serving=true, terminating=true) only if none are ready.
+	if len(eligibleSelectedEndpoints) == 0 {
+		for _, endpoint := range selectedEndpoints {
```

Comment:
So this is the third time you have iterated through endpoints but you could have done it once. In a single iteration, you can:

- apply the condition function, and if it passes the filter
- then apply the elegible function and accumulate in either slice 
- then you pick the proper slice and iterate once more with the transforming function

Reply:
Oh I see. Yes, let's avoid the extra iteration.